File: src/infrastructure/storage.shared/IAggregateStorage.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Threading.Tasks;
using core.fs.Accounts;
using core.Shared;

namespace storage.shared
{
    public interface IAggregateStorage
    {
        Task<IEnumerable<AggregateEvent>> GetEventsAsync(string entity, UserId userId);
        Task SaveEventsAsync(IAggregate agg, string entity, UserId userId, IDbTransaction outsideTransaction = null);
        Task SaveEventsAsync(IAggregate old, IAggregate newAggregate, string entity, UserId userId, IDbTransaction outsideTransaction = null);
        Task DoHealthCheck();
        Task DeleteAggregates(string entity, UserId userId, IDbTransaction outsideTransaction = null);
        Task DeleteAggregate(string entity, Guid aggregateId, UserId userId);
    }
}

File: src/infrastructure/storage.shared.fs/IAggregateStorage.fs
namespace storage.shared

open System
open System.Collections.Generic
open System.Data
open System.Threading.Tasks
open core.fs.Accounts
open core.Shared

type IAggregateStorage =
    abstract member GetEventsAsync : entity:string * userId:UserId -> Task<IEnumerable<AggregateEvent>>
    abstract member GetEventsAsync : entity:string * aggregateId:Guid * userId:UserId -> Task<IEnumerable<AggregateEvent>>
    abstract member SaveEventsAsync : agg:IAggregate * entity:string * userId:UserId * outsideTransaction:IDbTransaction -> Task
    abstract member SaveEventsAsync : old:IAggregate * newAggregate:IAggregate * entity:string * userId:UserId * outsideTransaction:IDbTransaction -> Task
    abstract member DoHealthCheck : unit -> Task
    abstract member DeleteAggregates : entity:string * userId:UserId * outsideTransaction:IDbTransaction -> Task
    abstract member DeleteAggregate : entity:string * aggregateId:Guid * userId:UserId -> Task
